Ravi, Xiong, and Howard<br>
CS 6140: Machine Learning<br>
Northeastern University<br>
March 27, 2023

## HW 6: Comprehensive Supervised Learning

### 0. Imports and data clean up

##### 0.1 Import libraries

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score, confusion_matrix, mean_squared_error, precision_score, recall_score, f1_score, r2_score
from sklearn.model_selection import train_test_split
from sklearn.neighbors import LocalOutlierFactor

# Deliver with style
plt.style.use('dark_background')
import warnings
warnings.filterwarnings('ignore') 
sns.set(rc={'figure.figsize':(11,8)})
pd.options.display.float_format = '{:.2f}'.format

##### 0.2 Read the data

In [2]:
# Read data from csv
df19 = pd.read_csv('data/brfss19.csv', usecols = range(1, 250))
df21 = pd.read_csv('data/brfss21.csv', usecols = range(1, 250))

# Shape of the dataframes
print(df19.shape)
print(df21.shape)

(418268, 249)
(438693, 249)


In [3]:
df19.head()

_STATE  FMONTH    IDATE  IMONTH  IDAY  IYEAR  DISPCODE       SEQNO  \
0       1       1  1182019       1    18   2019      1100  2019000001   
1       1       1  1132019       1    13   2019      1100  2019000002   
2       1       1  1182019       1    18   2019      1100  2019000003   
3       1       1  1182019       1    18   2019      1200  2019000004   
4       1       1  1042019       1     4   2019      1100  2019000005   

         _PSU  CTELENM1  ...  _FRUTSU1  _VEGESU1  _FRTLT1A  _VEGLT1A  _FRT16A  \
0  2019000001      1.00  ...    200.00    114.00         1         1        1   
1  2019000002      1.00  ...    100.00    121.00         1         1        1   
2  2019000003      1.00  ...    114.00    164.00         1         1        1   
3  2019000004      1.00  ...       NaN       NaN         9         9        1   
4  2019000005      1.00  ...    143.00    178.00         1         1        1   

   _VEG23A  _FRUITE1  _VEGETE1  _FLSHOT7  _PNEUMO3  
0        1         0         0      2.00      1.00  
1        1         0         0      1.00      1.00  
2        1         0         0      1.00      2.00  
3        1         1         1      9.00      9.00  
4        1         0         0      2.00      1.00  

[5 rows x 249 columns]

In [4]:
df21.head()

_STATE  FMONTH        IDATE IMONTH   IDAY    IYEAR  DISPCODE  \
0    1.00    1.00  b'01192021'  b'01'  b'19'  b'2021'   1100.00   
1    1.00    1.00  b'01212021'  b'01'  b'21'  b'2021'   1100.00   
2    1.00    1.00  b'01212021'  b'01'  b'21'  b'2021'   1100.00   
3    1.00    1.00  b'01172021'  b'01'  b'17'  b'2021'   1100.00   
4    1.00    1.00  b'01152021'  b'01'  b'15'  b'2021'   1100.00   

           SEQNO          _PSU  CTELENM1  ...  _MISVEG1  _FRTRES1  _VEGRES1  \
0  b'2021000001' 2021000001.00      1.00  ...      0.00      1.00      1.00   
1  b'2021000002' 2021000002.00      1.00  ...      0.00      1.00      1.00   
2  b'2021000003' 2021000003.00      1.00  ...      0.00      1.00      1.00   
3  b'2021000004' 2021000004.00      1.00  ...      0.00      1.00      1.00   
4  b'2021000005' 2021000005.00      1.00  ...      0.00      1.00      1.00   

   _FRUTSU1  _VEGESU1  _FRTLT1A  _VEGLT1A  _FRT16A  _VEG23A  _FRUITE1  
0    100.00    214.00      1.00      1.00     1.00     1.00      0.00  
1    100.00    128.00      1.00      1.00     1.00     1.00      0.00  
2    100.00     71.00      1.00      2.00     1.00     1.00      0.00  
3    114.00    165.00      1.00      1.00     1.00     1.00      0.00  
4    100.00    258.00      1.00      1.00     1.00     1.00      0.00  

[5 rows x 249 columns]

##### 0.3 Align the dataframes

In [5]:
# Compare the different column names in the two data frame
(set(df19.columns) | set(df21.columns)) - (set(df19.columns) & set(df21.columns))

{'CHCCOPD2', 'CHCCOPD3', 'INCOME2', 'INCOME3', '_AIDTST4', '_VEGETE1'}

* The dataframes diverge in only two response column names, which are 'CHCCOPD2' and 'CHCCOPD3'.
* Since the two names stand for the same thing, we will rename both of these columns df21['CHCCOPD3'] and df19['CHCCOPD2] as 'COPD'.
* For 'INCOME2' and 'INCOME3' we will also rename and preserve as 'INCOME2'.
* We will drop '_AIDTST4' and '_VEGETE1' entirely.

In [6]:
# Renaming the columns to match each other
df21.rename(columns = {'CHCCOPD3':'COPD'}, inplace = True)
df19.rename(columns = {'CHCCOPD2':'COPD'}, inplace = True)

In [7]:
# Renaming INCOME3 of df21 to INCOME2 to match with df19
df21.rename(columns = {'INCOME3' : 'INCOME2'}, inplace = True)

In [8]:
# Dropping _AIDTST4 and _VEGETE1
df21 = df21.drop('_AIDTST4', axis=1)
df19 = df19.drop('_VEGETE1', axis=1)

In [9]:
# Confirming no issue
print(set(df21.columns) - set(df19.columns))
for column in df21.columns:
    if column not in set(df19.columns):
        print(column)
for column in df19.columns:
    if column not in set(df21.columns):
        print(column)

set()


##### 0.4 Combine the dataframes

In [10]:
# Now the two df has the same column names, we can concat the two df.
df = pd.concat([df19, df21], ignore_index=True, sort=False)
df.head()

_STATE  FMONTH    IDATE IMONTH IDAY IYEAR  DISPCODE       SEQNO  \
0    1.00    1.00  1182019      1   18  2019   1100.00  2019000001   
1    1.00    1.00  1132019      1   13  2019   1100.00  2019000002   
2    1.00    1.00  1182019      1   18  2019   1100.00  2019000003   
3    1.00    1.00  1182019      1   18  2019   1200.00  2019000004   
4    1.00    1.00  1042019      1    4  2019   1100.00  2019000005   

           _PSU  CTELENM1  ...  _VEGRES1  _FRUTSU1  _VEGESU1  _FRTLT1A  \
0 2019000001.00      1.00  ...      1.00    200.00    114.00      1.00   
1 2019000002.00      1.00  ...      1.00    100.00    121.00      1.00   
2 2019000003.00      1.00  ...      1.00    114.00    164.00      1.00   
3 2019000004.00      1.00  ...      0.00       NaN       NaN      9.00   
4 2019000005.00      1.00  ...      1.00    143.00    178.00      1.00   

   _VEGLT1A  _FRT16A  _VEG23A  _FRUITE1  _FLSHOT7  _PNEUMO3  
0      1.00     1.00     1.00      0.00      2.00      1.00  
1      1.00     1.00     1.00      0.00      1.00      1.00  
2      1.00     1.00     1.00      0.00      1.00      2.00  
3      9.00     1.00     1.00      1.00      9.00      9.00  
4      1.00     1.00     1.00      0.00      2.00      1.00  

[5 rows x 248 columns]

### 1. Identifying the response variables and the feature space

Prepare the compound response variables:
* Depression: df['ADDEPEV3']
* Ever Told Asthma: df['_CASTHM1']
* Copd: df['COPD']
* Cancer: df['CHCSCNCR'], df['CHCOCNCR']
* Ever Told Heart Condition: df['CVDCRHD4'], df['CVDINFR4'], df['CVDSTRK3']
* Diabetes: df['DIABETE4']<br>

Then handle missing values and prepare the dataframe for analysis.<br>

##### 1.1 'Inclusive or' for cancers and cardiovascular diseases

In [11]:
# For cancer, we need to combine df['CHCSCNCR'] and f['CHCOCNCR']
print(df['CHCSCNCR'].unique())
print(df['CHCOCNCR'].unique())

[ 2.  1. nan  7.  9.]
[ 2.  1.  9.  7. nan]


Per the Codebook Report, only 1 stands for diagnosed. If either column shows 1, then the combined column is 1. Otherwise, we use 0 to stand for not diagnosed.

In [12]:
# Combine two different types of cancer into one feature
has_cncr = (df['CHCSCNCR'] == 1) | (df['CHCOCNCR'] == 1)
df['CNCR'] = has_cncr.astype(int)
df['CNCR'].value_counts()

0    708342
1    148619
Name: CNCR, dtype: int64

In [13]:
# Combine three different cardio-vascular condition into one feature
has_cvd = (df["CVDCRHD4"] == 1) | (df["CVDINFR4"] == 1) | (df["CVDSTRK3"] == 1) 
df["CVD"] = has_cvd.astype(int)
df["CVD"].value_counts() 

0    761194
1     95767
Name: CVD, dtype: int64

In [14]:
# print the percentage of unique values for cancer and heart condition
print(df['CNCR'].value_counts() / len(df['CNCR']))
print(df['CVD'].value_counts() / len(df['CVD']))

0   0.83
1   0.17
Name: CNCR, dtype: float64
0   0.89
1   0.11
Name: CVD, dtype: float64


Before creating the two additional categorical columns, we need to clean the NaN values in the chronic condition columns.

##### 1.2 Handling missing response values

In [15]:
chronic_df = df[['ADDEPEV3', '_CASTHM1', 'COPD', 'CHCSCNCR', 'CHCOCNCR', 'CNCR', 'CVDCRHD4', 'CVDINFR4', 'CVDSTRK3', 'CVD', 'DIABETE4']]
print(chronic_df.shape)

(856961, 11)


In [16]:
# Missing value analysis in the chronic_df
# list all the columns and number of missing values in that column
missing_cols = chronic_df.isna().sum().reset_index()
missing_cols

index   0
0   ADDEPEV3  13
1   _CASTHM1   0
2       COPD  11
3   CHCSCNCR  10
4   CHCOCNCR  12
5       CNCR   0
6   CVDCRHD4  10
7   CVDINFR4  12
8   CVDSTRK3  13
9        CVD   0
10  DIABETE4  12

Seeing that there are only few missing values in these columns (<0.1%), we simply drop the associated rows.

In [17]:
df.dropna(subset=['ADDEPEV3', '_CASTHM1', 'COPD', 'CHCSCNCR', 'CHCOCNCR', 'CVDCRHD4', 'CVDINFR4', 'CVDSTRK3', 'DIABETE4'], inplace=True)

##### 1.3 Handling missing feature values

In [18]:
chronic_conditions = ['ADDEPEV3', '_CASTHM1', 'COPD', 'CNCR', 'CVD', 
    'DIABETE4']

In [19]:
missing_values_cols = df.isna().sum()
missing_values_cols = missing_values_cols[missing_values_cols!=0].sort_values()

missing_values_cols.reset_index()

index       0
0    CHCKDNY2       2
1       _RACE       3
2    _RACEG21       3
3    _RACEGR3       3
4    CHECKUP1       6
..        ...     ...
183   COLGSEX  856873
184  COLGHOUS  856873
185   WORKCFS  856941
186   HAVECFS  856941
187   TOLDCFS  856941

[188 rows x 2 columns]

In [20]:
threshold = 0.2 * len(df)
threshold

171388.2

Since we already have the condition of trimming the data frame to 20 features, We will drop features entirely if we're missing their values in more than 20% of the rows.

In [21]:
drop_features = missing_values_cols[missing_values_cols > threshold].index
drop_features

Index(['_CHISPNC', 'SAFETIME', 'CELLFON5', 'PVTRESD3', 'CTELNUM1', 'CADULT1',
       'CSTATE1', 'HHADULT', 'CELLSEX', 'LANDLINE',
       ...
       'CSRVCTL2', 'PRIRHEPC', 'TRETHEPC', 'HAVEHEPC', 'MEDSHEPB', 'COLGSEX',
       'COLGHOUS', 'WORKCFS', 'HAVECFS', 'TOLDCFS'],
      dtype='object', length=132)

In [22]:
len(drop_features)

132

Dropping 132 features still leaves us with plenty of columns to work with.

In [23]:
# Dropping the features having more than 20% of missing values
df = df.drop(drop_features, axis = 1)

For the rest of the dataframe, we will impute the missing values with each columns' mean.

In [24]:
impute_features = missing_values_cols[missing_values_cols < threshold].index
impute_features

Index(['CHCKDNY2', '_RACE', '_RACEG21', '_RACEGR3', 'CHECKUP1', '_PRACE1',
       '_MRACE1', 'MENTHLTH', 'EDUCA', 'RENTHOM1', 'GENHLTH', 'PHYSHLTH',
       'MARITAL', 'CPDEMO1B', 'VETERAN3', 'QSTLANG', 'EMPLOY1', '_MICHD',
       'CHILDREN', '_URBSTAT', '_METSTAT', 'INCOME2', 'WEIGHT2', 'EXERANY2',
       'HEIGHT3', 'DEAF', 'BLIND', 'DECIDE', 'DIFFWALK', 'DIFFDRES',
       'DIFFALON', 'SMOKE100', 'USENOW3', 'ALCDAY5', 'HTM4', 'HTIN4',
       'FLUSHOT7', 'FRUIT2', 'PNEUVAC4', 'FRUITJU2', 'FVGREEN1', 'WTKG3',
       'HIVTST7', 'FRENCHF1', 'POTATOE1', 'VEGETAB2', 'FTJUDA2_', 'FRUTDA2_',
       '_BMI5', '_BMI5CAT', 'GRENDA1_', 'FRNCHDA_', 'POTADA1_', 'VEGEDA2_',
       '_FRUTSU1', '_VEGESU1'],
      dtype='object')

In [25]:
from sklearn.impute import SimpleImputer

mean_imputer = SimpleImputer(missing_values = np.nan, strategy = 'mean')

# Impute an empty value by replacing it with mean 
def impute_by_mean(col_name):
    df[col_name] = mean_imputer.fit_transform(df[col_name].values.reshape(-1, 1))


In [26]:
# Imputing missing values for features having < 20% missing values
for col in impute_features:
    impute_by_mean(col)

In [27]:
# Confirm no data is still missing
still_missing_values_cols = df.isna().sum()
still_missing_values_cols = still_missing_values_cols[still_missing_values_cols!=0].sort_values()
still_missing_values_cols.reset_index()

Empty DataFrame
Columns: [index, 0]
Index: []

##### 1.4 Creating our response columns for comorbidity

Create the binary categorical response column.

In [28]:
### 1.2 Create two additional categorical columns - CMB_binary, CMN_multi
has_chronic = (df['CNCR'] == 1)  | (df['ADDEPEV3'] == 1) | (df['_CASTHM1'] == 1) | (df['COPD'] == 1) | (df['CVD'] == 1) | (df['DIABETE4'] == 1)
df['CMB_binary'] = has_chronic.astype(int)
df['CMB_binary'].unique()

array([1, 0])

In [29]:
df["CMB_binary"].value_counts()

1    826641
0     30300
Name: CMB_binary, dtype: int64

Create the multiclass response column.

In [30]:
# Creating a subset of chronic conditions
cf = df[chronic_conditions]

In [31]:
# Replace values of 1 and 2 with 1 and rest with 0
cf["DIABETE4"] = cf["DIABETE4"].apply(lambda x: 1 if (x == 1) | (x == 2) else 0)

# Replace values of 1 with 1 and rest with 0 in other features
for col in chronic_conditions:
    cf[col] = cf[col].apply(lambda x: 1 if x == 1 else 0)

In [32]:
# Determine the number of chronic conditions an individual has
num_conditions = cf.sum(axis=1)

# Creating a new response column with the num of conditions
df['CMB_multiclass'] = pd.cut(num_conditions, bins=[0, 1, 2, 3, float('inf')], labels=[0, 1, 2, '3+'])


In [33]:
df["CMB_multiclass"].value_counts()

0     450760
1     250601
2      91764
3+     33916
Name: CMB_multiclass, dtype: int64

In [34]:
response_cols = ["CMB_binary", "CMB_multiclass"]

We now have response columns created and in a workable list. And, we can now avoid using the columns these responses were derived from in our feature space.

#### 1.5 Identifying our feature space

* 'INCOME2' = Annual household income from all sources
* '_AGE_G' = Six-level imputed age category
* 'MARITAL' = Marital status
* '_SEX'= Calculated sex variable
* '_EDUCAG' = Level of education completed
* '_RACE' = Race/ethnicity categories
* '_RFHYPE5' = Adults who have been told they have high blood pressure by a doctor, nurse, or other health professional
* '_HCVU651' = Respondents aged 18-64 who have any form of health care coverage
* '_MENT14D' = 3 level not good mental health status: 0 days, 1-13 days, 14-30 days
* '_PHYS14D' = 3 level not good physical health status: 0 days, 1-13 days, 14-30 days
* '_URBSTAT' = Urban/Rural status
* 'VETERAN3' = Have you ever served on active duty in the United States Armed Forces, either in the regular military or in a National Guard or military reserve unit?
* 'RENTHOM1' = Do you own or rent your home?
* 'TOLDHI2' = Have you ever been told by a doctor, nurse or other health professional that your blood cholesterol is high?
* 'CHECKUP1' = About how long has it been since you last visited a doctor for a routine checkup?  [A routine checkup is a general physical exam, not an exam for a specific injury, illness, or condition.]
* 'HAVARTH4' = (Ever told) (you had) some form of arthritis, rheumatoid arthritis, gout, lupus, or fibromyalgia?  (Arthritis diagnoses include: rheumatism, polymyalgia rheumatica; osteoarthritis (not osteporosis); tendonitis, bursitis, bunion, tennis elbow; carpal tunnel syndrome, tarsal tunnel syndrome; joint infection, etc.)
* 'CHCKDNY2' = Not including kidney stones, bladder infection or incontinence, were you ever told you had kidney disease?
* 'EXERANY2' = During the past month, other than your regular job, did you participate in any physical activities or exercises such as running, calisthenics, golf, gardening, or walking for exercise?
* 'STRENGTH' = During the past month, how many times per week or per month did you do physical activities or exercises to STRENGTHEN your muscles?   [Do NOT count aerobic activities like walking, running, or bicycling. Count activities using your own body weight like yoga, sit-ups or push-ups and those using weight machines, free weights, or elastic bands.]
* 'FRUIT2' = Not including juices, how often did you eat fruit?
* 'FVGREEN1' = How often did you eat a green leafy or lettuce salad, with or without other vegetables?
* 'FLUSHOT7' = During the past 12 months, have you had either flu vaccine that was sprayed in your nose or flu shot injected into your arm?<br>

We only need 20, but are starting with 22, so that we have a 10% buffer to delete problematic columns.

In [35]:
feature_cols = ['INCOME2', '_AGE_G', 'MARITAL', '_SEX', '_EDUCAG', '_RACE', '_RFHYPE5', '_HCVU651', '_MENT14D', '_PHYS14D', '_URBSTAT', 'VETERAN3', 'RENTHOM1', 'TOLDHI2', 'CHECKUP1', 'HAVARTH4', 'CHCKDNY2', 'EXERANY2', 'STRENGTH', 'FRUIT2', 'FVGREEN1', 'FLUSHOT7']

## Analysis Part 1:

### 2. Exploratory Data Analysis